File: blind-75/maximum_depth_bin_tree.java
class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;
	
	TreeNode(int val) {
		this.val = val;
	}
	
	TreeNode(int val, TreeNode left, TreeNode right) {
		this.val = val;
		this.left = left;
		this.right = right;
	}
}

class Solution {
	public int dfsToFindMaxDepth(TreeNode root) {
		if (root != null) {
			int leftDepth = dfsToFindMaxDepth(root.left);
			int rightDepth = dfsToFindMaxDepth(root.right);
			
			return 1 + Math.max(leftDepth, rightDepth);
		}
		
		return 0;
	}
	
	public int maxDepth(TreeNode root) {
		return dfsToFindMaxDepth(root);
	}
}

class Main {
	public static void main (String[] args) {
		Solution obj = new Solution();
		Test tester = new Test();
		
		// Test case 1
		TreeNode root = new TreeNode(
			1,
			new TreeNode(2),
			new TreeNode(3)
		);
		tester.assertEqual("Test case 1", obj.maxDepth(root), 2);
		
		// Test case 2
		root = new TreeNode(
			1,
			new TreeNode(2),
			new TreeNode(
				3,
				new TreeNode(4),
				new TreeNode(5)
			)
		);
		tester.assertEqual("Test case 1", obj.maxDepth(root), 3);
	}
}

File: blind-75/maximum_depth_bin_tree.py
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def maxDepth(self, root: TreeNode) -> int:
    
    def dfsToMaxDepth(root: TreeNode) -> int:
      if root:
        leftDepth = dfsToMaxDepth(root.left)
        rightDepth = dfsToMaxDepth(root.right)
        
        return 1 + max(leftDepth, rightDepth)
      
      return 0
    
    return dfsToMaxDepth(root)


obj = Solution()

def test__001():
  root = TreeNode(
    1,
    TreeNode(2),
    TreeNode(3)
  )
  assert(obj.maxDepth(root)) == 2

def test__002():
  root = TreeNode(
    1,
    TreeNode(2),
    TreeNode(
      3,
      TreeNode(4),
      TreeNode(5)
    )
  )
  assert(obj.maxDepth(root)) == 2